torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,257 @@
+
+def schemas_match(functional_schema, inplace_schema):
+    names_match = inplace_schema.name.endswith("_") and inplace_schema.name[:-1] == functional_schema.name
+    arg_types_match = all(
+        a1.type == a2.type for a1, a2 in zip(functional_schema.arguments, inplace_schema.arguments))
+    # for the inplace op, its first argument should be mutable
+    assert inplace_schema.arguments[0].alias_info is not None and inplace_schema.arguments[0].alias_info.is_write
+    # and its remaining arguments shouldn't be.
+    assert all(a.alias_info is None for a in inplace_schema.arguments[1:])
+    return names_match and arg_types_match
+
+def maybe_get_inplace_op(op):
+    # __module__ seems broken; it returns torch._ops.aten which doesn't exist
+    op_namespace = op.__module__.split(".")[-1]
+    op_base_name = op.overloadpacket.__name__
+    maybe_namespace_module = getattr(torch.ops, op_namespace)
+    maybe_inplace_op = None if maybe_namespace_module is None else getattr(maybe_namespace_module, f'{op_base_name}_', None)
+    if maybe_inplace_op is None:
+        return None
+
+    inplace_overloads = [
+        getattr(maybe_inplace_op, overload_name) for overload_name in maybe_inplace_op.overloads()
+    ]
+    inplace_overloads_with_matching_schemas = [
+        f
+        for f in inplace_overloads
+        if schemas_match(op._schema, f._schema)
+    ]
+    # This is for sanity: if foo() and foo_() are both operators,
+    # we expect them to have compatible schemas.
+    # (This is asserted by codegen for ATen, but might not be true
+    # for other arbitrary operators).
+    assert len(inplace_overloads_with_matching_schemas) == 1
+    inplace_op = inplace_overloads_with_matching_schemas[0]
+    return inplace_op
+
+@compatibility(is_backward_compatible=True)
+def reinplace(gm: fx.GraphModule, *sample_args, **sample_kwargs):
+    """
+    Given an fx.GraphModule, modifies it to perform "reinplacing",
+    mutating the nodes of the graph.
+    We look for out-of-place op call sites like `b = a.add(...)`,
+    and convert them to be inplace (`b = a.add_(...)`),
+    as long as the input to the current operator ("a") isn't re-used
+    anywhere later in the graph.
+
+    Sample inputs are needed to determine aliasing relationships of the inputs.
+    In general, we can't reinplace node `b = a.add(...)` if "a" aliases any of the
+    inputs to the program.
+
+    There is one exception though: if "a" is copied to at any point later in the program,
+    e.g. `a.copy_(...)`, then we are free to re-use a's buffer any time
+    before that node executes.
+
+    This is an important optimization to include, to ensure that after running
+    functionalization and then reinplacing, we don't regress the net memory usage
+    of the original model.
+    """
+
+
+    sample_args = [a.to(device='meta') if isinstance(a, torch.Tensor) else a for a in sample_args]
```

Comment:
OK so after the Friday discussion we agreed that we should be using the fake tensor stored metadata to get this information, rather than recomputing it on the fly in the Interpreter. https://github.com/pytorch/pytorch/pull/80566/files#diff-10df4b56573d38527c0d4485d31ac62c3aa4d76077f8d2066eb71abb7c2bc0c7R101-R128 is an example of how to do it in the new style but you'll need https://github.com/pytorch/pytorch/pull/80566/files#diff-342fa21508bcf89a6b85251ce90db2ea534a1bb06f72441a4714569f47d1d9c9R32-R42 landed somewhere first so you can actually populate these nodes (actually, we should probably make proxy tensor populate this as we are tracing; cc @Chillee ). Sorry about the API whiplash.

Btw if you still wanted to do this as is you should create a fake tensor mode and then use the converter there to convert to fake tensors, as that will maintain aliasing relationships in the inputs, which this code does not do correctly.